Port forwarding: preserve streamConn on error if any bytes sent/received

brianpursley opened the pull request:
#### What type of PR is this?

/kind bug
#### What this PR does / why we need it:

Currently, PortForwarder closes streamConn whenever an error occurs while handling a connection. However, this causes problems for users when a non-fatal error (like "connection reset by peer") has occurred because it causes PortForwarder to terminate when the remote end is still available and able to handle connections.

This change only closes streamConn when no bytes were sent or received on the connection before the error occurred.

While this is a general bug fix, it resolves an issue when using `kubectl port-forward` to connect to postgres, where postgres will sometimes return a "connection reset by peer" error during its normal operation, resulting in the connection being lost and kubectl exiting. With this change, the connection will be preserved allowing client tools like psql, datagrip, etc to function without interruption.

#### Which issue(s) this PR fixes:
Fixes #111825

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kubectl port-forward preserves port forwarding session when a non-fatal error occurs
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

ardaguclu commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -407,7 +410,11 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
-		pf.streamConn.Close()
+
+		// If no bytes were sent or received, the connection was never established and streamConn should be closed so that port forwarding can exit.
+		if bytesReceived == 0 && bytesSent == 0 {
```
This is good idea. What happens if 1 or more bytes are received or sent but we get [an error](https://github.com/kubernetes/kubernetes/pull/117442/files#diff-03eefd1f0ca3e8aaad80a0a95db19e19d9f092caed73a007645a03363d6676a3R412) because datastream is already closed https://github.com/kubernetes/kubernetes/pull/117442/files#diff-03eefd1f0ca3e8aaad80a0a95db19e19d9f092caed73a007645a03363d6676a3R393

brianpursley committed:
Port forwarding: preserve streamConn on error if any bytes sent/received

Currently, PortForwarder closes streamConn whenever an
error occurs while handling a connection. However, this
causes problems for users when a non-fatal error (like "connection reset by peer") has occurred because it causes PortForwarder to terminate
when the remote end is still available and able to handle
connections.

This change only closes streamConn when no bytes
were sent or received on the connection before the error
occurred.

fedebongio commented:
/assign @aojea 
Could you help with this one Antonio? Thank you!
/triage accepted

brianpursley commented:
> > However, this causes problems for users when a non-fatal error (like "connection reset by peer") has occurred because it causes PortForwarder to terminate when the remote end is still available and able to handle connections.
> 
> do you mean non-fatal or retriable, both things are different, connection reset by peer in network is fatal, means you tried to connect to something that is not listening.
> 

Thanks for the questions.

By non-fatal, I mean the handled connection should close (I think we can agree on that part), but the port forwarder should continue to handle other connections which have not been reset, and it should not stop listening for new connections, because the remote endpoint is still alive and able to handle new connections.

> I don't feel that the number of bytes can give us a good idea of what happens with the network though, since this are streamed bytes

My reasoning is that a connection that successfully sent or received any data indicates that the remote endpoint is alive, but it is just a heuristic for determining that.  Ideally I want to handle "connection reset by peer", but my attempts to solve the problem this way in a previous PR were not acceptable because it matched on the error string.

I am open to other suggestions.  In my last PR, it was suggested that a fix might be needed in the container runtime, to surface error details better, but that seems like a much more difficult way to solve this problem and something I am not confident would happen in a reasonable amount of time.

My goal is to fix the problem that kubectl users are experiencing, so with that in mind, can you think of any way to solve the issue?

Otherwise, I guess we might have to consider adding an option that enables the pre-1.23 behavior of never closing streamconn (even on error) but that has its own problems.  At least with an option, people can choose between two broken behaviors depending on their needs, but that would not be my first choice.

I realize this solution in this PR might not be "optimal" from a design standpoint, but practically, I believe it solves a real problem that people are having with kubectl port-forward.

brianpursley commented:
If this PR isn't a suitable approach, what do you think about reopening and reconsidering https://github.com/kubernetes/kubernetes/pull/111860?

I am looking for a way forward in this issue, one way or another, even if it is something different altogether that I haven't considered.  I am open to anything.

brianpursley commented:
@aojea it would be great if https://github.com/kubernetes/kubernetes/pull/117493 fixes this problem too, but I'm not sure it will.  It looks like that PR handles the EPIPE error, but in this case, the error coming out of the container is ECONNRESET.  I could be wrong, but I think it is probably similar, but a slightly different issue.

I pulled down the PRs for containerd and kubernetes, built them, and upgraded a kind cluster to use them and still had the same "connection reset by peer" kubectl exit problem, but maybe I did something wrong.

We can wait and see if those PRs fix this issue though, if you think they will.

aojea commented:
> I could be wrong, but I think it is probably similar, but a slightly different issue

It is in the other direction iiuic, first the piping is broken with the reset and when it tries to write gets the epipe error...

Anyway, before doing more forward fixing we need an e2e tests reproducing the issue, once we have that is simple to fix it

brianpursley commented:
> > I could be wrong, but I think it is probably similar, but a slightly different issue
> 
> It is in the other direction iiuic, first the piping is broken with the reset and when it tries to write gets the epipe error...
> 
> Anyway, before doing more forward fixing we need an e2e tests reproducing the issue, once we have that is simple to fix it

OK.  I have a reproduction now that doesn't involve postgres, just a simple TCP server written in C that can trigger RST on demand:
https://github.com/brianpursley/TestTcpServer

I'll see if I can turn this into an e2e test somehow...

aojea commented:
> I'll see if I can turn this into an e2e test somehow...

in golang if you send a curl to a TCP server like this and you immediately close the connection it sends a RST

```
package main

import (
	"fmt"
	"net"
	"os"
)

func main() {

	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		panic(err)
	}
	defer ln.Close()

	for {
		// Listen for an incoming connection.
		conn, err := ln.Accept()
		if err != nil {
			fmt.Println("Error accepting: ", err.Error())
			os.Exit(1)
		}
		// reset connection
		conn.Close()
	}

}
```
```
$ go run main.go
```
```
$ ss -pan | grep main
tcp   LISTEN     0      4096                                                                            127.0.0.1:40749                  0.0.0.0:*        users:(("main",pid=3044428,fd=3)) 

$ curl localhost:40749
curl: (56) Recv failure: Connection reset by peer

$ sudo tcpdump -nv -i any tcp port 40749
tcpdump: data link type LINUX_SLL2
tcpdump: listening on any, link-type LINUX_SLL2 (Linux cooked v2), snapshot length 262144 bytes
07:39:37.834264 lo    In  IP (tos 0x0, ttl 64, id 52270, offset 0, flags [DF], proto TCP (6), length 60)
    127.0.0.1.60960 > 127.0.0.1.40749: Flags [S], cksum 0xfe30 (incorrect -> 0xaafe), seq 2625484757, win 65495, options [mss 65495,sackOK,TS val 1295939107 ecr 0,nop,wscale 7], length 0
07:39:37.834273 lo    In  IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60)
    127.0.0.1.40749 > 127.0.0.1.60960: Flags [S.], cksum 0xfe30 (incorrect -> 0xefa6), seq 1626576637, ack 2625484758, win 65483, options [mss 65495,sackOK,TS val 1295939107 ecr 1295939107,nop,wscale 7], length 0
07:39:37.834282 lo    In  IP (tos 0x0, ttl 64, id 52271, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.60960 > 127.0.0.1.40749: Flags [.], cksum 0xfe28 (incorrect -> 0x1663), ack 1, win 512, options [nop,nop,TS val 1295939107 ecr 1295939107], length 0
07:39:37.834343 lo    In  IP (tos 0x0, ttl 64, id 52272, offset 0, flags [DF], proto TCP (6), length 131)
    127.0.0.1.60960 > 127.0.0.1.40749: Flags [P.], cksum 0xfe77 (incorrect -> 0x1ad3), seq 1:80, ack 1, win 512, options [nop,nop,TS val 1295939107 ecr 1295939107], length 79
07:39:37.834347 lo    In  IP (tos 0x0, ttl 64, id 144, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.40749 > 127.0.0.1.60960: Flags [.], cksum 0xfe28 (incorrect -> 0x1615), ack 80, win 511, options [nop,nop,TS val 1295939107 ecr 1295939107], length 0
07:39:37.834384 lo    In  IP (tos 0x0, ttl 64, id 145, offset 0, flags [DF], proto TCP (6), length 52)
    127.0.0.1.40749 > 127.0.0.1.60960: Flags [R.], cksum 0xfe28 (incorrect -> 0x1610), seq 1, ack 80, win 512, options [nop,nop,TS val 1295939107 ecr 1295939107], length 0

```

We can add it as one of the existing golang, I was thinking in adding a new handler here

https://github.com/kubernetes/kubernetes/blob/4c9411232e10168d7b050c49a1b59f6df9d7ea4b/test/images/agnhost/netexec/netexec.go#L230

I didn't test this code, but I think that this will have to generate a RST,  let me know if you want me to test it or just contact me in slack if you need help

````
mux.HandleFunc("/reset", resetConnectionHandler)
````

```
	http.HandleFunc("/reset", func(w http.ResponseWriter, r *http.Request) {
		hj, ok := w.(http.Hijacker)
		if !ok {
			http.Error(w, "webserver doesn't support hijacking", http.StatusInternalServerError)
			return
		}
		conn, _, err := hj.Hijack()
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
                conn.Close()
}
```
In another variation of a similar test, this is not exactly the same problem but based on the assumptions discussed I think the expectation is to be able to connect again, we have this:

1. run agnhost image to run a server 
```
kubectl run pf --image=registry.k8s.io/e2e-test-images/agnhost:2.39 --  netexec --http-port=8080
```
2. create port-forward
```
 kubectl port-forward  pods/pf 9090:8080
Forwarding from 127.0.0.1:9090 -> 8080
Forwarding from [::1]:9090 -> 8080
```
3. check it works (verify it returns the name of the pod, this is important to avoid false positives)
```
curl localhost:9090/hostname
pf
```
4. restart the server (agnhost exits on the special handler /exit)
```
 curl localhost:9090/exit
curl: (52) Empty reply from server
```
5. Try to connect again to the pod
```
curl localhost:9090/hostname
curl: (52) Empty reply from server
```
The port-forward will exist because it is trying to connect to the old container namespace
```
E0502 07:50:18.829820 3069980 portforward.go:409] an error occurred forwarding 9090 -> 8080: error forwarding port 8080 to pod 03322202c19dd46d4580085a7ce292c6750d939e14a9bf8f053d6faad26b88e9, uid : failed to execute portforward in network namespace "/var/run/netns/cni-7e321f87-4387-820f-d92a-ca4a4729070c": failed to connect to localhost:8080 inside namespace "03322202c19dd46d4580085a7ce292c6750d939e14a9bf8f053d6faad26b88e9", IPv4: dial tcp4 127.0.0.1:8080: connect: connection refused IPv6 dial tcp6 [::1]:8080: connect: connection refused
error: lost connection to pod
```

brianpursley commented:
> in golang if you send a curl to a TCP server like this and you immediately close the connection it sends a RST

Thank you for the info and suggestions.

While your example does indeed return connection reset by peer when running locally and curling as in your example, I put it in a pod and configured port forwarding and it did not result in the connection reset by peer error, but instead gets `curl: (52) Empty reply from server` and port forwarding keeps running, so it does not exhibit the problem described by this issue. :thinking: 

So far, aside from postgres+ssl, the only way I have been able to reproduce it is using [this C program](https://github.com/brianpursley/TestTcpServer/blob/main/tcp_server_rst.c) which explicitly sends RST.

```
$ kubectl run test2 --image=brianpursley/test-tcp-server:latest --image-pull-policy=Always
```

```
$ nc localhost 5000
```

```
$ kubectl port-forward test2 5000
Forwarding from 127.0.0.1:5000 -> 5000
Forwarding from [::1]:5000 -> 5000
Handling connection for 5000
E0502 20:06:07.555841   20742 portforward.go:406] an error occurred forwarding 5000 -> 5000: error forwarding port 5000 to pod 8ce90f66f1c0b690ae242d414c8402b1a76cdbce8eeafa1459fea7542dd2a264, uid : failed to execute portforward in network namespace "/var/run/netns/cni-572872e8-82df-8fef-387b-9a14657189a6": read tcp4 127.0.0.1:34062->127.0.0.1:5000: read: connection reset by peer
E0502 20:06:07.556049   20742 portforward.go:234] lost connection to pod
```

I'll keep looking into it to see if there is a way to turn this into an e2e test, but I am curious if you are able to reproduce "connection reset by peer" using a go program running in a pod, because I am not. 

(Incidentally, I read somewhere you cannot send an RST from a go program unless you use unsafe and do some low-level stuff)

aojea commented:
> (Incidentally, I read somewhere you cannot send an RST from a go program unless you use unsafe and do some low-level stuff)

hehe, we already do that low level stuff ;) 

https://github.com/kubernetes/kubernetes/blob/master/test/images/regression-issue-74839/main.go

> I'll keep looking into it to see if there is a way to turn this into an e2e test, but I am curious if you are able to reproduce "connection reset by peer" using a go program running in a pod, because I am not.

sure, I'll try to get there at the end of the week

aojea commented:
@brianpursley this one I pasted here https://github.com/kubernetes/kubernetes/pull/117442#issuecomment-1531062183 also generates a RST when it tries to reconnect, since there  is nothing on the port listening.

I added a new functionality to agnhost to do the reset by peer thing too https://github.com/kubernetes/kubernetes/pull/117764, so ideally we should have 2 e2e tests, one for connection refused and other for reset by peer


1. run agnhost image to run a server 
```
kubectl run pf --image=registry.k8s.io/e2e-test-images/agnhost:2.39 --  netexec --http-port=8080
```
2. create port-forward
```
 kubectl port-forward  pods/pf 9090:8080
Forwarding from 127.0.0.1:9090 -> 8080
Forwarding from [::1]:9090 -> 8080
```
3. check it works (verify it returns the name of the pod, this is important to avoid false positives)
```
curl localhost:9090/hostname
pf
```
4. restart the server (agnhost exits on the special handler /exit)
```
 curl localhost:9090/exit
curl: (52) Empty reply from server
```
5. Try to connect again to the pod
```
curl localhost:9090/hostname
curl: (52) Empty reply from server
```
The port-forward will exist because it is trying to connect to the old container namespace
```
E0502 07:50:18.829820 3069980 portforward.go:409] an error occurred forwarding 9090 -> 8080: error forwarding port 8080 to pod 03322202c19dd46d4580085a7ce292c6750d939e14a9bf8f053d6faad26b88e9, uid : failed to execute portforward in network namespace "/var/run/netns/cni-7e321f87-4387-820f-d92a-ca4a4729070c": failed to connect to localhost:8080 inside namespace "03322202c19dd46d4580085a7ce292c6750d939e14a9bf8f053d6faad26b88e9", IPv4: dial tcp4 127.0.0.1:8080: connect: connection refused IPv6 dial tcp6 [::1]:8080: connect: connection refused
error: lost connection to pod
```

aojea commented:
image ready

```
 kubectl run test --image=registry.k8s.io/e2e-test-images/agnhost:2.44 -- tcp-reset
```

e2e test can be added to reproduce the issue

brianpursley commented:
Thanks, @aojea 

I used the image you created to add an e2e test reproducing the issue.

https://github.com/kubernetes/kubernetes/pull/117851

brianpursley commented:
There is some other work being done currently to address this in container runtimes and kubelet that I think will address this problem in a more robust way.

pablochacin commented:
@brianpursley Any chance to re-open this PR? From what I can see it seems to address the main issue some users like me have, namely, terminating the port-forwarding on non-fatal cases, while https://github.com/kubernetes/kubernetes/issues/113568 has no clear path forward.

brianpursley commented:
> Hi @brianpursley any chance to re-open this PR? From what I understand, it should address the issue some users like me are experimenting, while the other ongoing efforts either may not address them at all [#117493](https://github.com/kubernetes/kubernetes/pull/117493) or are at un uncertain point [#113568](https://github.com/kubernetes/kubernetes/issues/113568).

Probably not, and the reason is this PR doesn't really fix the root problem, it just provides a workaround that is sort of a hack.  Understandably, the Kubernetes org wants to fix the root problem and avoid workaround fixes.

The PR to watch is https://github.com/kubernetes/kubernetes/pull/117493.  This fixes the problem at its root, as it is really a problem with the container runtime and how kubelet doesn't handle connection reset by peer.  It is working its way through the process, but there are several moving pieces and steps that need to happen, so it isn't moving as fast as everyone would like I'm sure.  It is making progress though and I am very hopeful (the corresponding required change to containerd has already been merged, which is a big hurdle crossed already).

Alternative solutions for you right now:

* Older versions of kubectl (I forget exactly which version offhand) might still work for port forwarding as ironically older versions had a different bug that just so happened to allow connection reset by peer to not cause the connection to be lost.  In more recent versions of kubectl that other bug was fixed, but as a side effect of that bug being fixed, this connection reset by peer problem emerged.

* There is a kubectl plugin called relay that you could try.  I haven't tried it, but it looks good and might avoid this problem: https://github.com/knight42/krelay

pablochacin commented:
> the Kubernetes org wants to fix the root problem and avoid workaround fixes.
Makes sense. Thanks.
